services/galley/src/Galley/External/LegalHoldService.hs

```diff
@@ -0,0 +1,228 @@
+import qualified Galley.Data.LegalHold      as LegalHoldData
+import qualified Data.ByteString.Lazy.Char8 as LC8
+import qualified Data.ByteString            as BS
+import qualified Network.HTTP.Client        as Http
+import qualified OpenSSL.EVP.Digest         as SSL
+import qualified OpenSSL.EVP.PKey           as SSL
+import qualified OpenSSL.PEM                as SSL
+import qualified OpenSSL.RSA                as SSL
+import qualified OpenSSL.Session            as SSL
+import qualified Ssl.Util                   as SSL
+import qualified System.Logger.Class        as Log
+
+----------------------------------------------------------------------
+-- api
+
+-- | Get /status from legal hold service; throw 'Wai.Error' if things go wrong.
+checkLegalHoldServiceStatus :: Fingerprint Rsa -> HttpsUrl -> Galley ()
+checkLegalHoldServiceStatus fpr url = do
+    resp <- makeVerifiedRequestFreshManager fpr url reqBuilder
+    if | Bilge.statusCode resp < 400 -> pure ()
+       | otherwise -> do
+            Log.info . Log.msg $ showResponse resp
+            throwM legalHoldServiceBadResponse
+  where
+    reqBuilder :: Http.Request -> Http.Request
+    reqBuilder
+        = Bilge.paths ["status"]
+        . Bilge.method GET
+        . Bilge.expect2xx
+
+-- | @POST /initiate@.
+requestNewDevice :: TeamId -> UserId -> Galley NewLegalHoldClient
+requestNewDevice tid uid = do
+    resp <- makeLegalHoldServiceRequest tid reqParams
+    case eitherDecode (responseBody resp) of
+        Left e -> do
+            Log.info . Log.msg $ "Error decoding NewLegalHoldClient: " <> e
+            throwM legalHoldServiceBadResponse
+        Right client -> pure client
+  where
+    reqParams =
+        Bilge.paths ["initiate"]
+      . Bilge.json (RequestNewLegalHoldClient uid tid)
+      . Bilge.method POST
+      . Bilge.acceptJson
+      . Bilge.expect2xx
+
+-- | @POST /confirm@
+-- Confirm that a device has been linked to a user and provide an authorization token
+confirmLegalHold :: ClientId
+                 -> TeamId
+                 -> UserId
+                 -> OpaqueAuthToken -- ^ TODO: Replace with 'LegalHold' token type
+                 -> Galley ()
+confirmLegalHold clientId tid uid legalHoldAuthToken = do
+    void $ makeLegalHoldServiceRequest tid reqParams
+  where
+    reqParams =
+        Bilge.paths ["confirm"]
+      . Bilge.json (LegalHoldServiceConfirm clientId uid tid (opaqueAuthTokenToText legalHoldAuthToken))
+      . Bilge.method POST
+      . Bilge.acceptJson
+      . Bilge.expect2xx
+
+-- | @POST /remove@
+-- Inform the LegalHold Service that a user's legalhold has been disabled.
+removeLegalHold :: TeamId
+                 -> UserId
+                 -> Galley ()
+removeLegalHold tid uid = do
+    void $ makeLegalHoldServiceRequest tid reqParams
+  where
+    reqParams =
+        -- TODO: Currently this OVERWRITES any path on the base URL;
+        -- We should come up with a better solution.
+        Bilge.paths ["remove"]
```

Comment:
We should definitely fix this before merging; this path should be appended to `legalHoldServiceUrl`

Reply:
see above